drv/cosmo-seq-server/src/vcore.rs

```diff
@@ -0,0 +1,466 @@
+
+        // Clear our faults
+        self.clear_faults(Rails {
+            vddcr_cpu0: true,
+            vddcr_cpu1: true,
+        })?;
+
+        // The higher-level sequencer code will unmask the FPGA interrupts for
+        // our guys.
+
+        ringbuf_entry!(Trace::Initialized);
+
+        Ok(())
+    }
+
+    pub fn clear_faults(&self, which_rails: Rails) -> Result<(), ResponseCode> {
+        if which_rails.vddcr_cpu0 {
+            retry_i2c_txn(Rail::VddcrCpu0, PmbusCmd::ClearFaults, || {
+                self.vddcr_cpu0.clear_faults()
+            })?;
+        }
+
+        if which_rails.vddcr_cpu1 {
+            retry_i2c_txn(Rail::VddcrCpu0, PmbusCmd::ClearFaults, || {
+                self.vddcr_cpu1.clear_faults()
+            })?;
+        }
+
+        ringbuf_entry!(Trace::FaultsCleared(which_rails));
+
+        Ok(())
+    }
+
+    pub fn handle_pmbus_alert(
+        &self,
+        mut rails: Rails,
+        now: u64,
+    ) -> InterruptAction {
+        ringbuf_entry!(Trace::PmbusAlert {
+            timestamp: now,
+            rails,
+        });
+
+        let mut should_sample_vin = false;
+        self.record_pmbus_status(
+            now,
+            Rail::VddcrCpu0,
+            &mut rails.vddcr_cpu0,
+            &mut should_sample_vin,
+        );
+
+        self.record_pmbus_status(
+            now,
+            Rail::VddcrCpu1,
+            &mut rails.vddcr_cpu1,
+            &mut should_sample_vin,
+        );
+
+        if should_sample_vin {
+            for _ in 0..VCORE_NSAMPLES {
+                let vddcr_cpu0_vin =
+                    self.vddcr_cpu0.read_vin().unwrap_or_else(|e| {
+                        // We don't retry I2C errors here, since we're just going
+                        // to take another reading anyway.
+                        ringbuf_entry!(Trace::I2cError(
+                            Rail::VddcrCpu0,
+                            PmbusCmd::ReadVin,
+                            e
+                        ));
+                        units::Volts(f32::NAN)
+                    });
+
+                let vddcr_cpu1_vin =
+                    self.vddcr_cpu1.read_vin().unwrap_or_else(|e| {
+                        // We don't retry I2C errors here, since we're just going
+                        // to take another reading anyway.
+                        ringbuf_entry!(Trace::I2cError(
+                            Rail::VddcrCpu1,
+                            PmbusCmd::ReadVin,
+                            e
+                        ));
+                        units::Volts(f32::NAN)
+                    });
+                //
+                // Record our readings, along with a timestamp.  On the
+                // one hand, it's a little exceesive to record a
+                // timestamp on every reading:  it's in milliseconds,
+                // and because it takes ~900µs per reading, we expect
+                // the timestamp to (basically) be incremented by 2 with
```

Comment:
Why is this "incremented by 2"?  I would expect it to be incremented by 1, since ticks ≈ ms

Reply:
we take two readings, one from each VRM.